src/opustrainer/modifiers/placeholders.py

```diff
@@ -287,54 +283,100 @@ def __call__(self, line:str) -> str:
+                    alignments[:candidate_index]
+                    # fill in the gap created by the replaced bit: the 2-1 2-2 2-3 bit
+                    + [Pair(candidate.src + src_tag_offset, candidate.trg + n) for n in range(len(augment_tokens))]
+                    # fill in the gap created by the replaced bit: the 4-1 5-2 6-3 bit
+                    + [Pair(candidate.src + trg_tag_offset + n, candidate.trg + n) for n in range(len(augment_tokens))]
+                    # pairs after the replaced bit have to be offset by the length of the replacement
+                    # bit minus the length of the bit we replaced (1)
+                    + [Pair(pair.src + len(tag_tokens) - 1, pair.trg) for pair in alignments[candidate_index+1:]]
+                )
+                candidate_offset = candidate_index + 2 * len(augment_tokens)
+
             elif mode == "augment":
                 # Augment mode adds random noise both on the source and the target without any
                 # tagging encouraging the model to copy crap from one side to the other.
-                augment = get_random_unicode_string()
-                source[candidates[i][0]] = source[candidates[i][0]] + " " + augment
-                target[candidates[i][1]] = target[candidates[i][1]] + " " + augment
-
-        source_detok = self.src_detokenizer.detokenize(source)
-        target_detok = self.trg_detokenizer.detokenize(target)
+                augment_tokens = get_random_unicode_strings()
+                source = source[:candidate.src+1] + augment_tokens + source[candidate.src+1:]
+                target = target[:candidate.trg+1] + augment_tokens + target[candidate.trg+1:]
+
+                # Fix up alignment pairs
+                alignments = (
+                    # pairs before and including the candidate stay the same
+                    alignments[:candidate_index+1]
+                    # fill in the gap created by the added random noise
+                    + [Pair(candidate.src + n, candidate.trg + n) for n in range(1, len(augment_tokens) + 1)]
+                    # pairs after the replaced bit have to be offset by the length of the replacement bit
+                    + [Pair(pair.src + len(augment_tokens), pair.trg + len(augment_tokens)) for pair in alignments[candidate_index+1:]]
+                )
+                candidate_offset = candidate_index + len(augment_tokens) + 1
+
+            else:
+                raise RuntimeError('unknown mode')
+
+        source_detok, _, source_mapping = self.src_retokenizer.retokenize(source)
+        target_detok, _, target_mapping = self.trg_retokenizer.retokenize(target)
+        remapped_pairs = compute_mapping(source_mapping, target_mapping, alignments)
 
         # Return the sentence, source tagged a la Dinu et al, target as it is and no alignment info
-        return  source_detok + "\t" + target_detok
+        return  source_detok + "\t" + target_detok + "\t" + format_alignments(remapped_pairs)
```

Comment:
In our case, I'd say that if you don't supply an spm_vocab, you don't care about the alignments because Marian only really should ever get alignments that are aligned to the spm tokens. But that's a very us-case.

Then again, at this moment, its only us that uses this code anyway. And it already assumes Marian with the whole detokized-but-spm-alignments-plz input. We can care about other trainers in the future or via #29.

Okay if I turn this line into a `+ ("\t" + format_alignments(...) if spm_vocab else '')` sort of thing?

Reply:
It now only produces alignments when `spm_vocab` is passed in.